### PR TITLE
Fleet Autoscaler custom sync: Race condition fix

### LINKED ISSUE #2385: Flakiness: Autoscaler tests
**What happened**:

Autoscaler tests have a tendency to be flaky.

Usually a `AssertFleetCondition` fails.

**What you expected to happen**:

Not be flaky 😄 

**How to reproduce it (as minimally and precisely as possible)**:

Usually by running CI.

You might be able to get it to run via running lots of the e2e tests over and over. I've found `TestAutoscalerStressCreate` is often a decent candidate.

**Anything else we need to know?**:

I have a two current theories I need to confirm:

1. The flakiness is in when the `CustomFasSyncInterval` is enabled, but I wasn't checking on this before when tracking the issues, so need to go back and look.
2. At least part of the problem lies in the fact that the `fasThreads` map in the autoscaler controller was using the FleetAutoscaler `name` for the key. This has no guarantee to be unique, and could therefore cause race conditions (thinking removal comes after a new addition with the same name, which happens in some e2e tests). We should use the UID of the object instead to guarantee uniqueness of the key.

**Environment**: Cloud Build
- Agones version:  dev.
- Kubernetes version (use `kubectl version`): 1.21
- Cloud provider or hardware configuration: GKE
- Install method (yaml/helm): helm
- Troubleshooting guide log(s): N/A
- Others:

Digging into some fixes here: 
https://github.com/markmandel/agones/tree/wip/autoscaler-flakiness

### PR DESCRIPTION
**Pull request by markmandel**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, press enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind breaking

/kind bug

> /kind cleanup
> /kind documentation
> /kind feature
> /kind hotfix

**What this PR does / Why we need it**:

Re-architected how the polling and autoscaling works such that it's more
inline with the original implementation:

* Switch from sync.Map to using a mutex, as it allows the code to be transactional when doing multiple operations on the fasThread map.
* Switch out using the terminate channel into using a cancellable context. This removes the possibility of deadlocks on an already closed channel (issue in previous implementation).
* Move the sync loop to use `wait.Until()`, since it's already tested and utilised.
* Make the loop only send an enqueue message rather than fire the autoscale logic. This gives us back retry logic on failure, and also means there are no errors that can be returned in that process that could abort the syncing routing (which was in the previous implementation).
* Moved to using the ObjectMeta.UID rather than name for the map key, as we have no guarantee that the name is unique between runs (and some e2e tests reused the same name, and caused race conditions).

Also includes some cleanup and extra debugging in tests as I worked through what the originating issue was.

**Which issue(s) this PR fixes**:

Closes #2385

**Special notes for your reviewer**:

🤞🏻

### CODE REVIEW

**Comment by markmandel** [comment]
@jie-bao if you are interested, would love to get your take on this - since you wrote the initial implementation.

**Comment by markmandel** [comment]
That's a new one on me:

```
--- FAIL: TestFleetRollingUpdate (64.60s)
    --- FAIL: TestFleetRollingUpdate/Use_fleet_Patch_true_25% (2.78s)
        fleet_test.go:246: 
            	Error Trace:	fleet_test.go:246
            	Error:      	Expected nil, but got: &errors.StatusError{ErrStatus:v1.Status{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ListMeta:v1.ListMeta{SelfLink:"", ResourceVersion:"", Continue:"", RemainingItemCount:(*int64)(nil)}, Status:"Failure", Message:"Internal error occurred: failed calling webhook \"mutations.agones.dev\": Post \"https://agones-controller-service.agones-system.svc:443/mutate?timeout=10s\": dial tcp 10.40.2.228:8081: connect: connection refused", Reason:"InternalError", Details:(*v1.StatusDetails)(0xc000a0e780), Code:500}}
            	Test:       	TestFleetRollingUpdate/Use_fleet_Patch_true_25%
        fleet_test.go:250: 
            	Error Trace:	fleet_test.go:250
            	Error:      	Not equal: 
            	            	expected: 1
            	            	actual  : 0
            	Test:       	TestFleetRollingUpdate/Use_fleet_Patch_true_25%
```

**Comment by markmandel** [comment]
Legit bug, caused a couple of panics.

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x60 pc=0x139b93a]

goroutine 409 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc002939ba0})
	/go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:55 +0xd8
panic({0x1807f00, 0x2ac6f40})
	/usr/local/go/src/runtime/panic.go:1038 +0x215
agones.dev/agones/pkg/fleetautoscalers.(*Controller).deleteFasThread(0xc0000b3a40, 0x0, 0x1)
	/go/src/agones.dev/agones/pkg/fleetautoscalers/controller.go:462 +0xfa
agones.dev/agones/pkg/fleetautoscalers.(*Controller).syncFleetAutoscaler(0xc0000b3a40, {0x1ce3208, 0xc0005f0d00}, {0xc002966a20, 0x28})
	/go/src/agones.dev/agones/pkg/fleetautoscalers/controller.go:264 +0x796
agones.dev/agones/pkg/util/workerqueue.(*WorkerQueue).processNextWorkItem(0xc0004add60, {0x1ce3208, 0xc0005f0d00})
	/go/src/agones.dev/agones/pkg/util/workerqueue/workerqueue.go:180 +0x28e
agones.dev/agones/pkg/util/workerqueue.(*WorkerQueue).runWorker(...)
	/go/src/agones.dev/agones/pkg/util/workerqueue/workerqueue.go:156
agones.dev/agones/pkg/util/workerqueue.(*WorkerQueue).run.func1()
	/go/src/agones.dev/agones/pkg/util/workerqueue/workerqueue.go:215 +0x45
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x7fa1548fcaf0)
	/go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:155 +0x67
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x11c9d56, {0x1cb0dc0, 0xc000b00840}, 0x1, 0xc00009b320)
	/go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:156 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x111db8b, 0x3b9aca00, 0x0, 0x0, 0x0)
	/go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x89
k8s.io/apimachinery/pkg/util/wait.Until(...)
	/go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:90
agones.dev/agones/pkg/util/workerqueue.(*WorkerQueue).run(0xc0004add60, {0x1ce3208, 0xc0005f0d00})
	/go/src/agones.dev/agones/pkg/util/workerqueue/workerqueue.go:215 +0xdc
created by agones.dev/agones/pkg/util/workerqueue.(*WorkerQueue).Run
	/go/src/agones.dev/agones/pkg/util/workerqueue/workerqueue.go:204 +0x168
```

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x60 pc=0x139b93a]

goroutine 540 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc001f7d740})
	/go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:55 +0xd8
panic({0x1807f00, 0x2ac6f40})
	/usr/local/go/src/runtime/panic.go:1038 +0x215
agones.dev/agones/pkg/fleetautoscalers.(*Controller).deleteFasThread(0xc0002cef00, 0x0, 0x1)
	/go/src/agones.dev/agones/pkg/fleetautoscalers/controller.go:462 +0xfa
agones.dev/agones/pkg/fleetautoscalers.(*Controller).syncFleetAutoscaler(0xc0002cef00, {0x1ce3208, 0xc0005c38c0}, {0xc0016fd410, 0x21})
	/go/src/agones.dev/agones/pkg/fleetautoscalers/controller.go:264 +0x796
agones.dev/agones/pkg/util/workerqueue.(*WorkerQueue).processNextWorkItem(0xc0000994f0, {0x1ce3208, 0xc0005c38c0})
	/go/src/agones.dev/agones/pkg/util/workerqueue/workerqueue.go:180 +0x28e
agones.dev/agones/pkg/util/workerqueue.(*WorkerQueue).runWorker(...)
	/go/src/agones.dev/agones/pkg/util/workerqueue/workerqueue.go:156
agones.dev/agones/pkg/util/workerqueue.(*WorkerQueue).run.func1()
	/go/src/agones.dev/agones/pkg/util/workerqueue/workerqueue.go:215 +0x45
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x7fc1ee427fd0)
	/go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:155 +0x67
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x11c9d56, {0x1cb0dc0, 0xc0007eeb40}, 0x1, 0xc0006a5320)
	/go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:156 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000da57d0, 0x3b9aca00, 0x0, 0x78, 0xc00097eb60)
	/go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x89
k8s.io/apimachinery/pkg/util/wait.Until(...)
	/go/src/agones.dev/agones/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:90
agones.dev/agones/pkg/util/workerqueue.(*WorkerQueue).run(0xc0000994f0, {0x1ce3208, 0xc0005c38c0})
	/go/src/agones.dev/agones/pkg/util/workerqueue/workerqueue.go:215 +0xdc
created by agones.dev/agones/pkg/util/workerqueue.(*WorkerQueue).Run
	/go/src/agones.dev/agones/pkg/util/workerqueue/workerqueue.go:204 +0x168
```

**Comment by markmandel** [comment]
Both e2e tests passed, just got caught on the 404s in the site at the moment, which will be fixed in #2421

**Comment by markmandel** [comment]
```
--- FAIL: TestFleetRecreateGameServers (0.00s)
    --- FAIL: TestFleetRecreateGameServers/gameserver_shutdown (123.29s)
        fleet_test.go:1160: Could not message GameServer: timed out attempting to send UDP packet to address: timed out waiting for the condition
```

on e2e-stable

Different flake! That UDP connection flake is a weird one as well. I feel like it only happened on K8s upgrade 🤔

**Comment by markmandel** [comment]
Apologies if it wasn't clear - this should be good for review now 👍🏻

File: pkg/fleetautoscalers/controller.go [hunk 1]
```diff
@@ -377,71 +392,111 @@ func (c *Controller) updateStatusUnableToScale(ctx context.Context, fas *autosca
...
+		c.addFasThread(fas, false)
 	}
 }
 
-// syncFleetAutoscalerWithCustomSyncInterval syncs the fleet based on a custom interval configured in Fas
-func (c *Controller) syncFleetAutoscalerWithCustomSyncInterval(ctx context.Context, key string) error {
-	fas, err := c.getFleetAutoscalerByKey(key)
-	if fas == nil || err != nil {
-		return err
+// deleteFasThread remove a FleetAutoScaler sync routine
```
Comment (roberthbailey) [inline pkg/fleetautoscalers/controller.go:454]: nit: s/remove/removes

File: pkg/fleetautoscalers/controller.go [hunk 2]
```diff
@@ -377,71 +392,111 @@ func (c *Controller) updateStatusUnableToScale(ctx context.Context, fas *autosca
...
-	ticker := c.clock.NewTicker(time.Duration(fas.Spec.Sync.FixedInterval.Seconds) * time.Second)
-	c.fasThreads.Store(key, fasThread{
-		terminateSignal: make(chan struct{}),
-		generation:      fas.Generation,
-	})
-	// scale immediately when an FAS is created or updated
-	if err := c.fleetAutoScale(ctx, key); err != nil {
-		return err
+
+	if lock {
```
[the remark this answers is not in the document]
Comment (roberthbailey) [reply pkg/fleetautoscalers/controller.go:409]: Discussed offline.

**Review by roberthbailey — APPROVED** [review]

**Pull request merged**